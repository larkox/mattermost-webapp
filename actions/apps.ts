// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Client4} from 'mattermost-redux/client';
import {submitInteractiveDialog as submitInteractiveDialogRedux} from 'mattermost-redux/actions/integrations';
import {DialogSubmission} from 'mattermost-redux/types/integrations';
import {getCurrentChannelId} from 'mattermost-redux/selectors/entities/common';
import {ActionFunc, DispatchFunc, GetStateFunc} from 'mattermost-redux/types/actions';
import {AppCallResponseTypes, AppCallResponse, AppCall} from 'mattermost-redux/types/apps';

import {sendEphemeralPost} from 'actions/global_actions';
import {openInteractiveDialog} from 'plugins/interactive_dialog';
import {executeCommand} from 'actions/command';

export function doAppCall(call: AppCall): ActionFunc {
    return async (dispatch: DispatchFunc) => {
        const res = await Client4.executeAppCall(call) as AppCallResponse<any>; // TODO use unknown and check types (may need type guards)

        switch (res.type) {
        case AppCallResponseTypes.OK:
            if (res.markdown) {
                if (!call.context.channel_id) {
                    return {data: res};
                }

                sendEphemeralPost(res.markdown, call.context.channel_id, call.context.root_id);
            }
            return {data: res};
        case AppCallResponseTypes.MODAL:
            if (!res.data) {
                return {data: res};
            }
            res.data.app_id = call.context.app_id;
            openInteractiveDialog(res.data);
            return {data: res};
        case AppCallResponseTypes.COMMAND:
            res.data.args.command = res.data.command;
            return dispatch(executeCommand(res.data.command, res.data.args));
        }

        return {data: res};
    };
}

export function submitInteractiveDialog(dialog: DialogSubmission): ActionFunc {
    return (dispatch: DispatchFunc, getState: GetStateFunc) => {
        if (!dialog.submission.app_id) {
            return dispatch(submitInteractiveDialogRedux(dialog));
        }

        // Should `from` be used here instead of getting the channelID this way?
        // This interactive dialog could have been triggered by a command run in the RHS for example.
        // The previous Call or CallResponse would contain the correct channelID
        const channelID = getCurrentChannelId(getState());
        return dispatch(doAppCall({
            url: dialog.url,
            values: dialog.submission,
            context: {
                app_id: dialog.submission.app_id,
                channel_id: channelID,
            },
        }));
    };
}
