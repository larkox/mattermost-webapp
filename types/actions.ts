// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {ActionResult, GenericAction} from 'mattermost-redux/types/actions';
import { BatchAction } from 'redux-batched-actions';
import { GlobalState } from './store';

export type SuccessResult = {data: any;}

export type ErrorResult = {error: any;}

export function isSuccess(result: ActionResult): result is SuccessResult {
    return 'data' in result && !isError(result);
}

export function isError(result: ActionResult): result is ErrorResult {
    return Boolean((result as ErrorResult).error);
}
