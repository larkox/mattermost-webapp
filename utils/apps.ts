// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {AppBinding} from 'mattermost-redux/types/apps';

export function fillBindingsInformation(binding: AppBinding) {
    binding.bindings?.forEach((b) => {
        // Propagate id down if not defined
        if (!b.app_id) {
            b.app_id = binding.app_id;
        }

        // Compose location
        b.location = binding.location + '/' + b.location;

        // Propagate call down if not defined
        if (!b.call) {
            b.call = binding.call;
        }

        fillBindingsInformation(b);
    });

    // Trim branches without app_id
    if (!binding.app_id) {
        binding.bindings = binding.bindings?.filter((v) => v.app_id);
    }

    // Trim branches without calls
    if (!binding.call) {
        binding.bindings = binding.bindings?.filter((v) => v.call);
    }

    // Pull up app_id if needed
    if (binding.bindings?.length && !binding.app_id) {
        binding.app_id = binding.bindings[0].app_id;
    }

    // Pull up call if needed
    if (binding.bindings?.length && !binding.call) {
        binding.call = binding.bindings[0].call;
    }
}
