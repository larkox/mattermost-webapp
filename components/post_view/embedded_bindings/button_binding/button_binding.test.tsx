// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {Post} from 'mattermost-redux/types/posts';
import {AppBinding} from 'mattermost-redux/types/apps';

import {shallowWithIntl} from 'tests/helpers/intl-test-helper';

import ButtonBinding from './button_binding';

describe('components/post_view/embedded_bindings/button_binding/', () => {
    const post = {
        id: 'post_id',
        channel_id: 'channel_id',
    } as Post;

    const binding = {
        label: 'some_label',
        location: 'some_location',
        call: {
            path: 'some_url',
        },
    } as AppBinding;
    const baseProps = {
        post,
        userId: 'user_id',
        binding,
        actions: {
            doAppCall: jest.fn(async () => {
                return {data: {type: 'ok'}};
            }),
            getChannel: jest.fn(async (id: string) => {
                return {data: {id, team_id: 'team_id'}};
            }),
        },
    };

    test('should match snapshot', () => {
        const wrapper = shallowWithIntl(<ButtonBinding {...baseProps}/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('should call handleAction on click', (done) => {
        const doAppCall = jest.fn(async () => {
            done();
            return {data: {type: 'ok'}};
        });

        const props = {
            ...baseProps,
            actions: {
                ...baseProps.actions,
                doAppCall,
            },
        };

        const wrapper = shallowWithIntl(<ButtonBinding {...props}/>);

        wrapper.find('button').simulate('click');
    });
});
