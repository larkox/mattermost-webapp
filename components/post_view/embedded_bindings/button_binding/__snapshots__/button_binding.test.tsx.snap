// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/post_view/embedded_bindings/button_binding/ should match snapshot 1`] = `
<button
  onClick={[Function]}
>
  <LoadingWrapper
    loading={false}
    text={null}
  >
    <Connect(Markdown)
      options={
        Object {
          "autolinkedUrlSchemes": Array [],
          "markdown": false,
          "mentionHighlight": false,
        }
      }
    />
  </LoadingWrapper>
</button>
`;
